0026-an-extensible-db-design.adoc

```diff
@@ -0,0 +1,181 @@
+. Certain data of a swap may only be known to a specific role at a later point in time (after creation of the swap). We want to avoid table updates.
+
+
+== Data collection
+
+We identified the following data sources needed:
+
+. Initial data of each swap when being created for a specific role for a specific alpha and beta to protocol where alpha and beta is one of: `HAN`, `HErc20`, `HALight`
+. Additional data added to the swap for a specific role (e.g. secret hash only known to Bob during announcement)
+. Additional data added to the swap's alpha or beta ledger data.
+
+While not necessarily in scope we considered the following as well:
+
+. Swap State (one per swap, aggregated state - record all states with timestamp)
+. Communication State (one per swap - record all states with timestamp)
+. Ledger State (two per swap, record all states with timestamp and additional information if applicable, e.g. transaction ID)
+
+
+== Final model
+
+Instead of starting from scratch, we start iterating over a slightly adapted table layout based on the one proposed by @bonomat
+https://3.basecamp.com/4403044/buckets/16118249/messages/2583518881#__recording_2600021433[in basecamp].
+
+The final model can be found at the end of this document.
+Our reasoning for creating certain table relations is summed up in the sections below.
+
+=== Swaps
+
+image::http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/14e942f9c5ea1791f01bb65db0e09ac11cf5ab4c/assets/db_schema/swap.puml&fmt=svg[Swaps]
+
+Our first challange was depicting `alpha` and `beta` protocols for the swap.
+Our first attempt was to not indicate this in the `swap` table but instead just use foreign key relationships.
+For that to work we would need to indicate in the child table whether it is `alpha` or `beta` (see below in `swap_protocol`).
+In order to reduce the amount of duplicated data we decided to add a separate table `protocol` and a `1 to many` relation ship from `swap` to `protocol`.
+`protocol` is a _static_ table.
+Static means, it contains all available protocols as identifiers.
+For now: `han_bitcoin`, `han_ethereum`, `herc20`, `halight`.
+
+This allows us to :
+
+* simple update the name if we decide to, i.e. instead of updating X rows in `swaps` we update 1 entry in `protocol`
+* add additional protocols
+
+The second thing which was a bit challenging was how to model data which is only known at a later point in time and how to model data which not all `Swaps` have in common.
+We decided to go for an _extension table_ approach which allows us to add additional information where needed:
+
+* Bob only learns the hash of the secret during the announcement, additionally not every Swap has a secret hash.
+* The shared swap id is only available after a swap was finalized.
+
+We added `secret_hash` and `shared_swap_id` as extension tables, that have the `swap_local_id` as primary key (so there can only be one entry per swap).
```

Comment:
I would recommend to use a `UNIQUE` constraint instead and let the database use IDs for whatever it needs.

Reply:
Sounds good and I think it's similar to what Tobin proposed.